sql/open_address_hash.h

```diff
@@ -0,0 +1,315 @@
+#pragma once
+
+#include <string.h>
+#include <my_dbug.h>
+
+
+
+template <typename trait, typename key_trait> class open_address_hash
+{
+public:
+  using T= typename trait::elem_type;
+  using find_type= typename trait::find_type;
+  using erase_type= typename trait::erase_type;
+  using hash_value_type= typename key_trait::hash_value_type;
+  using key_type= typename key_trait::key_type;
+
+  //key_type *get_key(const T &elem) { return key_trait::get_key(elem); }
+  bool is_empty(const T &el) { return trait::is_empty(el); }
+  void set_null(T &el) { trait::set_null(el); }
+
+  open_address_hash()
+  {
+    first.set_mark(true);
+    first.set_ptr(nullptr);
+    second= nullptr;
+  }     
+
+  ~open_address_hash()
+  {
+    if (!first.mark())
+    {
+      DBUG_ASSERT(hash_array);
+      free(hash_array);
+    }
+  }
+
+private:
+  hash_value_type to_index(const hash_value_type &hash_value)
+  {
+    return hash_value & (capacity - 1);
+  }
+
+  hash_value_type hash_from_value(const T &value)
+  {
+    return trait::get_hash_value(value);
+  }
+
+  bool insert_helper(const T &value)
+  {
+
+    auto hash_val= to_index(hash_from_value(value));
+
+    while (hash_array[hash_val] != nullptr)
+    {
+      if (hash_array[hash_val] == value)
+        return false;
+      hash_val= to_index(hash_val + 1);
+    }
+
+    hash_array[hash_val]= value;
+    _size++;
+    return true;
+  };
+
+  uint rehash_subsequence(uint i)
+  {
+    for (uint j= to_index(i + 1); hash_array[j] != nullptr; j= to_index(j + 1))
+    {
+      auto temp_el= hash_array[j];
+      hash_array[j]= nullptr;
+      //_size--;
+      insert_helper(temp_el);
```

Comment:
Sorry, initially I wanted to add the comment here, but sent it to the wrong `nullptr+insert_helper` pair:) But it als applies there.

Now here, concider we have this collision in memory:
```
|5|6|7|8|9|
|0|A|B|C|0|
```
And we're removing A, i.e. `hash_array[6]`. Now what happens in memory
```
1. hash_array[6] gets 0
2. hash_array[6] gets B
3. hash_array[7] gets 0
4. hash_array[7] gets C
5. hash_array[8] gets 0
```
Steps 1 and 3 are obsolete. If you'll find a way to avoid it, you'll make twice fewer stores during the collision rehashing!

Reply:
I understand what you are saying, but I have no idea how to implement it